tls/s2n_ktls_io.c

```diff
@@ -357,6 +398,7 @@ ssize_t s2n_ktls_sendv_with_offset(struct s2n_connection *conn, const struct iov
     size_t bytes_written = 0;
     POSIX_GUARD_RESULT(s2n_ktls_sendmsg(conn->send_io_context, TLS_APPLICATION_DATA,
             bufs, count, blocked, &bytes_written));
+    POSIX_GUARD_RESULT(s2n_ktls_enforce_estimated_record_limit(conn, bytes_written));
```

Comment:
Discussed offline and added a comment. We can't check before sending for s2n_sendfile, and I choose to keep the behavior consistent with s2n_send, but maybe folks disagree with that choice. We should error immediately because it's a fatal error, not a blocking error, and we can't assume the method will be called again.

Reply:
Don't we know the intended send size of sendfile with the `count` argument? Could we check if `count` is larger than the estimated remaining bytes before sending?

Reply:
`count` is just an upper bound, so there's no guarantee we'll actually send `count` bytes. But it's an interesting idea 🤔 We'd basically be putting the responsibility on applications not to make a call that could ever result in hitting the encryption limit, even if in practice the call wouldn't actually result in sending that many bytes.

I think I like that better than flagging the violation afterwards.

Reply:
Right. It seems like if `count` is larger than the estimated remaining bytes we're okay to fail, because their intent is clearly to send more than their limit even if those bytes would never be sent.

I wonder if we could also do something like
```
count = MIN(count, estimated_bytes_remaining);
```
and then we only fail when the limit is actually reached.

Reply:
I think that's adding too much complexity. We never expect this situation to occur; it only happens if the application has violated the contract. We should make the check as simple as possible.